src/Icicle/Dictionary/Parse/Types.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Icicle.Dictionary.Parse.Types where
+
+import qualified Data.HashMap.Strict as M
+import           Data.Int            (Int64)
+import           Data.List           (intersect)
+import           Data.Text           (Text)
+import qualified Data.Text           as T
+import           Data.Time.Clock     (UTCTime)
+import           Data.Time.Format    ()
+
+import qualified Text.Parsec.Pos as Pos
+
+-- | The 'Table' is a mapping ('HashMap') of 'Text' keys to 'Node' values.
+type Table = M.HashMap Text (Node, Pos.SourcePos)
+
+-- | A 'Node' may contain a 'TValue', or a 'Table'.
+data Node = NTValue TValue
+          | NTable  Table
+  deriving (Eq, Show)
+
+-- | A 'TValue' may contain any type of value that can put in a 'VArray'.
+data TValue = VString   [(Char, Pos.SourcePos)]
+            | VInteger  Int64
+            | VFloat    Double
+            | VBoolean  Bool
+            | VDatetime UTCTime
+            | VArray    [TValue]
+  deriving (Eq, Show)
+
+-- | Contruct an empty 'Table'.
+emptyTable :: Table
+emptyTable = M.empty
+
+
+-- | Contruct an empty 'NTable'.
+emptyNTable :: Node
+emptyNTable = NTable M.empty
+
+
+-- | Inserts a table ('Table') with name ('[Text]') which may be part of
+-- a table array (when 'Bool' is 'True') into a 'Table'.
+-- It may result in an error ('Text') on the 'Left' or a modified table
+-- on the 'Right'.
+insert :: ([Text], (Node, Pos.SourcePos)) -> Table -> Either Text Table
+insert ([], _)           _    = error "FATAL: Cannot call 'insert' without a name."
+insert (_ , (NTValue _, _)) _ = error "FATAL: Cannot call 'insert' with a TValue."
```

Comment:
yeah it should be doing it at the parse stage though, as early as possible.

don't get me wrong what @HuwCampbell is suggesting is OK, given that he is modifying/migrating some pre-existing base, but if this was being designed from the ground up, it would be better to get all of that out of the way in the parsing stage which has an `EitherT` specifically for "malformed input" errors.  once its parsed you want to be dealing wiht malformed input as little as possible